pkg/api/dirs_test.go

```diff
@@ -0,0 +1,247 @@
+					name:      "2.png",
+					path:      "img",
+					reference: swarm.MustParseHexAddress("b234ea7954cab7b2ccc5e07fe8487e932df11b2275db6b55afcbb7bad0be73fb"),
+					headers: http.Header{
+						"Content-Type": {"image/png"},
+					},
+				},
+			},
+		},
+	} {
+		t.Run("valid tar", func(t *testing.T) {
+			// create and collect all files in the test case
+			dirFiles := make([]*os.File, len(tc.files))
+			for i, file := range tc.files {
+				// create dir if the file is nested
+				if file.path != "" {
+					if _, err := os.Stat(file.path); os.IsNotExist(err) {
+						if err := os.Mkdir(file.path, 0700); err != nil {
+							t.Fatal(err)
+						}
+						defer os.RemoveAll(file.path)
+					}
+				}
+				f := writeAndOpenFile(t, path.Join(file.path, file.name), file.data)
+				defer os.Remove(f.Name())
+				defer f.Close()
+				dirFiles[i] = f
+			}
+
+			// tar all the test case files
+			tarReader := tarFiles(t, dirFiles)
+
+			// verify directory tar upload response
+			jsonhttptest.ResponseDirectSendHeadersAndReceiveHeaders(t, client, http.MethodPost, dirUploadResource, tarReader, http.StatusOK, api.FileUploadResponse{
+				Reference: swarm.MustParseHexAddress(tc.expectedHash),
+			}, nil)
+
+			// create expected manifest
+			expectedManifest := jsonmanifest.NewManifest()
+			for _, file := range tc.files {
+				e := &jsonmanifest.JSONEntry{
+					Reference: file.reference,
+					Name:      file.name,
+					Headers:   file.headers,
+				}
+				expectedManifest.Add(path.Join(file.path, file.name), e)
+			}
+
+			b, err := expectedManifest.Serialize()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// verify directory upload manifest through files api
+			jsonhttptest.ResponseDirectCheckBinaryResponse(t, client, http.MethodGet, fileDownloadResource(tc.expectedHash), nil, http.StatusOK, b, nil)
+		})
+	}
+}
+
+// writeAndOpenFile creates a new file with the given name and data and returns it as a variable ready for reading
+// callers should make sure to close and delete the created file
+func writeAndOpenFile(t *testing.T, name string, data []byte) *os.File {
+	err := ioutil.WriteFile(name, data, 0755)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	f, err := os.Open(name)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return f
+}
+
+// tarFiles receives an array of files and creates a new tar with those files as a collection
+// it returns a bytes.Buffer which can be used to read the created tar
+func tarFiles(t *testing.T, files []*os.File) *bytes.Buffer {
+	var buf bytes.Buffer
+	tw := tar.NewWriter(&buf)
+
+	for _, file := range files {
+		// get file info
+		info, err := file.Stat()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// create tar header and write it
+		hdr := &tar.Header{
+			Name: file.Name(),
+			Mode: 0600,
+			Size: info.Size(),
+		}
+		if err := tw.WriteHeader(hdr); err != nil {
+			t.Fatal(err)
+		}
+
+		// open and read the file data
+		r, err := os.Open(file.Name())
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer r.Close()
```

Comment:
not good. this will be called only when the entire function returns. please close this explicitly after `tw.Write` line

Reply:
what if `tw.Write` outputs and error and this `defer` is not in place?

Reply:
then call `r.Close()` inside the `err!=nil` block before `Fatal()`?

Reply:
whoops i missed something, why do you need to open the file again? aren't you already passing an open file handle when calling `tarFiles`??

Reply:
> then call r.Close() inside the err!=nil block before Fatal()?

**in general**, i can't see how having 2 `Close()` calls for the same file is better than leaving 1 `defer` inside a `for`.

can you please explain why it is a problem that deferred calls are invoked when the entire function returns? isn't that the whole point of `defer`?

i see this as a good thing; this way we know these statements will not slip by the cracks.

> whoops i missed something, why do you need to open the file again? aren't you already passing an open file handle when calling `tarFiles`??

you're right. i removed the redundant `Open()` in `tarFiles`.

--- 

leaving this aside, i am happy to integrate `dir_test.go` lines `154` to `169` into the `tarFiles` aux func if you think all the file & dir opening/closing logic is better contained this way.

the more i think about it, the more i am leaning towards this alternative.

please lmk your thoughts on this

Reply:
ok, looking at [this comment](https://github.com/ethersphere/bee/pull/447#discussion_r461166611) you made down below.

will update soon